openproblems/tasks/spatial_decomposition/methods/cell2location.py

```diff
@@ -0,0 +1,112 @@
+from ....tools.decorators import method
+from ....tools.utils import check_version
+from ..utils import split_sc_and_sp
+
+import numpy as np
+
+
+@method(
+    method_name="Cell2location",
+    paper_name="Cell2location maps fine-grained cell types in spatial transcriptomics",
+    paper_url="https://www.nature.com/articles/s41587-021-01139-4",
+    paper_year=2022,
+    code_url="https://github.com/BayraktarLab/cell2location",
+    image="openproblems-python-extras",
+)
+def stereoscope(adata, test=False, max_epochs_sc=None, max_epochs_sp=None):
+
+    from cell2location.models import Cell2location
+    from cell2location.models import RegressionModel
+
+    if test:
+        max_epochs_sp = max_epochs_sp or 10
+        max_epochs_sc = max_epochs_sc or 10
+        num_samples = 10
+    else:  # pragma: nocover
+        max_epochs_sc = max_epochs_sc or 250
+        max_epochs_sp = max_epochs_sp or 30000
+        num_samples = 1000
+
+    adata_sc, adata = split_sc_and_sp(adata)
+
+    if "tech" in adata_sc.obs.columns:
+        adata_sc.obs["batch_key"] = adata_sc.obs["tech"].copy()
+    else:
+        adata_sc.obs["batch_key"] = "all"
+
+    # REFERENCE SIGNATURE ESTIMATION FROM SCRNA
+    # prepare anndata for the regression model
+    RegressionModel.setup_anndata(
+        adata=adata_sc,
+        # 10X reaction / sample / batch
+        batch_key="batch_key",
+        # cell type, covariate used for constructing signatures
+        labels_key="label",
+    )
+    sc_model = RegressionModel(adata_sc)
+    sc_model.train(max_epochs=max_epochs_sc)
+    # In this section, we export the estimated cell abundance
+    # (summary of the posterior distribution).
+    adata_sc = sc_model.export_posterior(
+        adata_sc, sample_kwargs={"num_samples": num_samples, "batch_size": 2500}
+    )
+    # export estimated expression in each cluster
+    if "means_per_cluster_mu_fg" in adata_sc.varm.keys():
```

Comment:
I don't quite understand this logic. Why would this be present vs absent?

Reply:
Saving to .varm fails for some datasets for reasons that I don't understand (index mismatch issue between varm.index and var_names).

Reply:
When that fails, .var works

Reply:
To simplify, can we do
```
try:
    means_per_cluster = adata_sc.varm["means_per_cluster_mu_fg"]
except KeyError:
    # sometimes varm fails for unknown reason
    means_per_cluster = adata_sc.var

inf_aver = means_per_cluster[
            [
                f"means_per_cluster_mu_fg_{i}"
                for i in adata_sc.uns["mod"]["factor_names"]
            ]
        ].copy()
```
?